File: ios/HybridSegmentationResult.swift
import Foundation
import CoreGraphics
import ImageIO
import UniformTypeIdentifiers
import NitroModules

final class HybridSegmentationResult: HybridSegmentationResultSpec {
  private var rgba: Data
  private var mask: Data
  private let pixelWidth: Int
  private let pixelHeight: Int

  init(output: SegmentationOutput) {
    self.rgba = output.rgba
    self.pixelWidth = output.width
    self.pixelHeight = output.height
    self.bounds = output.bounds
    self.sourceWidth = Double(output.sourceWidth)
    self.sourceHeight = Double(output.sourceHeight)
    self.foregroundCoverage = output.foregroundCoverage
    self.centroid = output.centroid
    self.instanceCount = output.instanceCount
    self.pixelBounds = output.pixelBounds
    self.trimOrigin = output.trimOrigin
    self.mask = output.mask
    self.hasMask = output.hasMask
    super.init()
  }

  let bounds: Rect
  let sourceWidth: Double
  let sourceHeight: Double
  let foregroundCoverage: Double
  let centroid: NormalizedPoint
  let instanceCount: Double
  let pixelBounds: PixelRect
  let trimOrigin: NormalizedPoint
  let hasMask: Bool

  var width: Double { Double(pixelWidth) }
  var height: Double { Double(pixelHeight) }

  var memorySize: Int {
    rgba.count + mask.count + HybridMemorySize.overhead
  }

  func dispose() {
    rgba = Data()
    mask = Data()
  }

  func toMaskBuffer() throws -> Promise<ArrayBuffer> {
    guard hasMask else {
      throw RuntimeError("No mask retained. Pass retainMask: true to removeBackground.")
    }
    let maskCopy = mask
    return Promise.parallel(VisionKitQueue.queue) {
      try ArrayBuffer.copy(data: maskCopy)
    }
  }

  func toArrayBuffer() throws -> Promise<ArrayBuffer> {
    let rgbaCopy = rgba
    return Promise.parallel(VisionKitQueue.queue) {
      try ArrayBuffer.copy(data: rgbaCopy)
    }
  }

  func saveToTemporaryFile(format: ImageFormat, quality: Double) throws -> Promise<String> {
    let qualityClamped = min(100, max(0, Int(quality.rounded())))
    let width = pixelWidth
    let height = pixelHeight
    let rgbaCopy = rgba
    let formatCopy = format
    return Promise.parallel(VisionKitQueue.queue) {
      guard let cgImage = RgbaImageConversion.cgImage(fromPremultipliedRgba: rgbaCopy, width: width, height: height) else {
        throw RuntimeError("Failed to build image from masked pixels.")
      }
      let ext = formatCopy == .png ? "png" : "jpg"
      let utType = formatCopy == .png ? UTType.png : UTType.jpeg
      let url = FileManager.default.temporaryDirectory
        .appendingPathComponent("visionkit-\(UUID().uuidString).\(ext)")
      guard let destination = CGImageDestinationCreateWithURL(url as CFURL, utType.identifier as CFString, 1, nil) else {
        throw RuntimeError("Failed to create image destination.")
      }
      let options: [CFString: Any] = [
        kCGImageDestinationLossyCompressionQuality: Double(qualityClamped) / 100.0,
      ]
      CGImageDestinationAddImage(destination, cgImage, options as CFDictionary)
      guard CGImageDestinationFinalize(destination) else {
        throw RuntimeError("Failed to encode image.")
      }
      return url.path
    }
  }
}

File: android/src/main/java/com/margelo/nitro/nitrovisionkit/HybridSegmentationResult.kt
package com.margelo.nitro.nitrovisionkit

import android.graphics.Bitmap
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import java.io.File
import java.io.FileOutputStream

@Keep
@DoNotStrip
class HybridSegmentationResult internal constructor(
  output: SegmentationOutput,
  private val tempDir: File,
) : HybridSegmentationResultSpec() {

  private var rgba = output.pixels
  private var mask = output.mask
  private val pixelWidth = output.width
  private val pixelHeight = output.height

  override val bounds: Rect = output.bounds
  override val sourceWidth: Double = output.sourceWidth.toDouble()
  override val sourceHeight: Double = output.sourceHeight.toDouble()
  override val foregroundCoverage: Double = output.foregroundCoverage
  override val centroid: NormalizedPoint = output.centroid
  override val instanceCount: Double = output.instanceCount
  override val pixelBounds: PixelRect = output.pixelBounds
  override val trimOrigin: NormalizedPoint = output.trimOrigin
  override val hasMask: Boolean = output.hasMask

  override val width: Double
    get() = pixelWidth.toDouble()

  override val height: Double
    get() = pixelHeight.toDouble()

  override val memorySize: Long
    get() = rgba.size.toLong() + mask.size.toLong() + HybridMemorySize.OVERHEAD

  override fun dispose() {
    rgba = ByteArray(0)
    mask = ByteArray(0)
  }

  override fun toMaskBuffer(): Promise<ArrayBuffer> {
    if (!hasMask) {
      return Promise.rejected(
        RuntimeException("No mask retained. Pass retainMask: true to removeBackground."),
      )
    }
    val maskCopy = mask
    return Promise.async {
      ArrayBuffer.copy(maskCopy)
    }
  }

  override fun toArrayBuffer(): Promise<ArrayBuffer> {
    val rgbaCopy = rgba
    return Promise.async {
      ArrayBuffer.copy(rgbaCopy)
    }
  }

  override fun saveToTemporaryFile(format: ImageFormat, quality: Double): Promise<String> {
    val q = quality.toInt().coerceIn(0, 100)
    val w = width.toInt()
    val h = height.toInt()
    val pixels = rgba
    return Promise.async {
      val bitmap = RgbaConversions.premultipliedRgbaToBitmap(pixels, w, h)
      try {
        val compressed = if (format == ImageFormat.PNG) {
          Bitmap.CompressFormat.PNG
        } else {
          Bitmap.CompressFormat.JPEG
        }
        val ext = if (format == ImageFormat.PNG) "png" else "jpg"
        val file = File.createTempFile("visionkit-", ".$ext", tempDir)
        FileOutputStream(file).use { out -> bitmap.compress(compressed, q, out) }
        file.absolutePath
      } finally {
        bitmap.recycle()
      }
    }
  }
}
